### PR TITLE
[MRG] Jcpot : Multi source DA with target shift

### PR DESCRIPTION
**Pull request by ievred**
Added jcpot class in the da.py, jcpot_barycenter with the optimization routine, an example and a test

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -29,6 +29,7 @@ It provides the following solvers:
 * Non regularized free support Wasserstein barycenters [20].
 * Unbalanced OT with KL relaxation distance and barycenter [10, 25].
 * Screening Sinkhorn Algorithm for OT [26].
+* JCPOT algorithm for multi-source target shift [27].
```
Comment (rflamary) [inline README.md:32]: Multi source domain adaptation with target shift

File: ot/bregman.py [hunk 2]
```diff
@@ -1502,6 +1503,164 @@ def unmix(a, D, M, M0, h0, reg, reg0, alpha, numItermax=1000,
...
+            nbelemperclass = np.sum(Ys[d] == c)
+            if nbelemperclass != 0:
+                D1[int(c), Ys[d] == c] = 1.
+                D2[int(c), Ys[d] == c] = 1. / (nbelemperclass)
+        dom['D1'] = D1
+        dom['D2'] = D2
+
+        # build the cost matrix and the Gibbs kernel
+        M = dist(Xs[d], Xt, metric=metric)
+        M = M / np.median(M)
```
Comment (rflamary) [inline ot/bregman.py:1607]: normalization by median should be optional.

File: README.md (hunk 1 above)
Comment (ievred) [reply README.md]: done

File: ot/bregman.py (hunk 2 above)
Comment (ievred) [reply ot/bregman.py]: removed median and adjusted parameters in the tests accordingly

File: ot/bregman.py [hunk 3]
```diff
@@ -1502,6 +1503,165 @@ def unmix(a, D, M, M0, h0, reg, reg0, alpha, numItermax=1000,
...
+    - :math:`W_{reg}(\cdot,\cdot)` is the entropic regularized Wasserstein distance (see ot.bregman.sinkhorn)
+    - :math:`\mathbf{D}_2^{(k)}` is a matrix of weights related to k-th source domain defined as in [p. 5, 27], its expected shape is `(n_k, C)` where `n_k` is the number of elements in the k-th source domain and `C` is the number of classes
+    - :math:`\mathbf{h}` is a vector of estimated proportions in the target domain of size C
+    - :math:`\mathbf{a}` is a uniform vector of weights in the target domain of size `n`
+    - :math:`\mathbf{D}_1^{(k)}` is a matrix of class assignments defined as in [p. 5, 27], its expected shape is `(n_k, C)`
+
+    The problem consist in solving a Wasserstein barycenter problem to estimate the proportions :math:`\mathbf{h}` in the target domain.
+
+    The algorithm used for solving the problem is the Iterative Bregman projections algorithm
+    with two sets of marginal constraints related to the unknown vector :math:`\mathbf{h}` and uniform tarhet distribution.
```
[the remark this answers is not in the document]
Comment (ievred) [reply ot/bregman.py]: corrected

File: test/test_da.py [hunk 4]
```diff
@@ -549,3 +547,57 @@ def test_linear_mapping_class():
     Cst = np.cov(Xst.T)
 
     np.testing.assert_allclose(Ct, Cst, rtol=1e-2, atol=1e-2)
+
+
+def test_jcpot_transport_class():
```
[the remark this answers is not in the document]
Comment (ievred) [reply test/test_da.py:552]: done

File: ot/da.py [hunk 5]
```diff
@@ -1914,3 +1906,180 @@ class label
                 self.log_ = dict()
 
         return self
+
+
+class JCPOTTransport(BaseTransport):
+
+    """Domain Adapatation OT method for multi-source target shift based on Wasserstein barycenter algorithm.
+
```
[the remark this answers is not in the document]
Comment (ievred) [reply ot/da.py:1996]: added label prop for base class and jcpot + tests for all otda methods

[the remark this answers is not in the document]
Comment (rflamary) [reply ot/da.py:1996]: very nice!

File: ot/bregman.py [hunk 6]
```diff
@@ -1502,6 +1503,165 @@ def unmix(a, D, M, M0, h0, reg, reg0, alpha, numItermax=1000,
...
+            log['err'].append(err)
+
+        if verbose:
+            if cpt % 200 == 0:
+                print('{:5s}|{:12s}'.format('It.', 'Err') + '\n' + '-' * 19)
+                print('{:5d}|{:8e}|'.format(cpt, err))
+
+    bary = bary / np.sum(bary)
+
+    if log:
```
Comment (rflamary) [inline ot/bregman.py:1655]: Hum, do we really need to return the gammas by default?

The name of the function is barycenter which is the weights h in this case. 
 I would put the gammas in the log or at least second position.

File: ot/da.py [hunk 7]
```diff
@@ -941,6 +946,50 @@ class label
...
+            # set nans to 0
+            transp[~ np.isfinite(transp)] = 0
+
+            for c in classes:
+                D1[int(c), ysTemp == c] = 1
+
+            # compute transported samples
+            transp_ys = np.dot(D1, transp)
+
+            return np.argmax(transp_ys, axis=0)
```
Comment (rflamary) [inline ot/da.py:991]: why return argmax? I woul return the smooth label estimations and let the user do the argmax if he really wants one label.

File: ot/da.py [hunk 8]
```diff
@@ -1009,8 +1057,44 @@ class label
...
+            # set nans to 0
+            transp[~ np.isfinite(transp)] = 0
+
+            for c in classes:
+                D1[int(c), ytTemp == c] = 1
+
+            # compute transported samples
+            transp_ys = np.dot(D1, transp.T)
+
+            return np.argmax(transp_ys, axis=0)
```
Comment (rflamary) [inline ot/da.py:1093]: same comment

File: ot/da.py [hunk 9]
```diff
@@ -1914,3 +1988,260 @@ class label
...
+            for i in range(len(self.xs_)):
+
+                # perform label propagation
+                transp = self.coupling_[i] / np.sum(self.coupling_[i], 1)[:, None]
+
+                # set nans to 0
+                transp[~ np.isfinite(transp)] = 0
+
+                # compute transported labels
+                transp_ys.append(np.argmax(np.dot(D1, transp.T), axis=0))
```
Comment (rflamary) [inline ot/da.py:2245]: same argmax comment

File: ot/bregman.py (hunk 6 above)
Comment (ievred) [reply ot/bregman.py:1655]: as far as I understand we have to return gamma as in documentation of the BaseTransport of da.py it is said that: "fit method should estimate a coupling matrix and store it in a coupling_ attribute"

File: ot/da.py (hunk 9 above)
Comment (ievred) [reply ot/da.py]: Done, changed documentation to "soft labels"

File: ot/da.py (hunk 8 above)
Comment (ievred) [reply ot/da.py]: Done, changed documentation to "soft labels"

File: ot/da.py (hunk 7 above)
Comment (ievred) [reply ot/da.py]: Done, changed documentation to "soft labels"

File: ot/bregman.py (hunk 6 above)
Comment (ievred) [reply ot/bregman.py:1653]: done

**Pull request merged**